src/puya/puya-binary.ts

```diff
@@ -0,0 +1,210 @@
+import * as crypto from 'crypto'
+import * as fs from 'fs'
+import { fileURLToPath } from 'node:url'
+import * as path from 'path'
+import * as tar from 'tar'
+import upath from 'upath'
+import { Constants } from '../constants'
+import { InternalError } from '../errors'
+import { logger } from '../logger'
+import { generateTempDir } from '../util/generate-temp-file'
+import type { SemVer } from './semver'
+
+/**
+ * Gets the platform-specific binary name
+ * @returns The appropriate binary name for the current platform
+ */
+function getBinaryName(): string {
+  return process.platform === 'win32' ? 'puya.exe' : 'puya'
+}
+
+/**
+ * Finds the path to the node_modules directory
+ * @returns The absolute path to the node_modules directory
+ */
+function findNodeModulesDir(): string {
+  // Start with the current working directory
+  const moduleDir = upath.dirname(fileURLToPath(import.meta.url))
+  let currentDir = moduleDir
+
+  // Keep going up directories until we find node_modules or hit the root
+  while (currentDir !== path.parse(currentDir).root) {
+    const potentialNodeModulesDir = path.join(currentDir, 'node_modules')
+    if (fs.existsSync(potentialNodeModulesDir)) {
+      return potentialNodeModulesDir
+    }
+    currentDir = path.dirname(currentDir)
+  }
+
+  // If we couldn't find it, default to the current directory
+  return moduleDir
+}
+
+function getPuyaStorageDir(version: SemVer): string {
+  const nodeModulesDir = findNodeModulesDir()
+  return path.join(nodeModulesDir, '.puya', version.formatted)
+}
+
+/**
+ * Finds the path to a cached Puya binary if it exists
+ * @returns The path to the cached binary or undefined if not found
+ */
+export function findCachedPuyaBinary(version: SemVer): string | undefined {
+  const puyaStorageDir = getPuyaStorageDir(version)
```

Comment:
I'm wondering if there is another way to do this, rather than searching for the node modules and the binary each time compile is requested.

Reply:
Sorry just saw your other comment.